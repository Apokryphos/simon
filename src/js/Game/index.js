function getRandomInt(min, max) {
  return min + Math.floor(Math.random() * max);
}

function Game(buttonCount = 8, random = null) {
  this.strict = false;

  this.getRandomButton = random || (() => getRandomInt(0, this.buttonCount));

  this.reset(buttonCount);
}

Game.prototype.getButtonCount = function getButtonCount() {
  return this.buttonCount;
};

Game.prototype.getSequence = function getSequence() {
  return this.sequence.slice();
};

Game.prototype.getSequenceLength = function getSequenceLength() {
  return this.sequence.length;
};

Game.prototype.getStrict = function getStrict() {
  return this.strict;
};

Game.prototype.reset = function reset(buttonCount) {
  this.buttonCount = buttonCount || this.buttonCount;
  this.sequenceIndex = 0;
  this.sequence = [this.getRandomButton()];
};

Game.prototype.setStrict = function setStrict(strict) {
  this.strict = strict;
};

Game.prototype.submit = function submit(buttonIndex) {
  if (buttonIndex >= 0 && buttonIndex < this.buttonCount) {
    if (this.sequence[this.sequenceIndex] === buttonIndex) {
      ++this.sequenceIndex;

      if (this.sequenceIndex >= this.sequence.length) {
        this.sequence.push(this.getRandomButton());
        this.sequenceIndex = 0;
      }

      return true;
    }

    if (this.strict) {
      this.reset();
    } else {
      this.sequenceIndex = 0;
    }

    return false;
  }

  throw new Error('Button index out of range.');
};

export default Game;
