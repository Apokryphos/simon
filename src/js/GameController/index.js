import EventEmitter from '../EventEmitter';

const GameState = {
  READY: 0,
  BUSY: 1,
};

function GameController(game, playDelay) {
  this.playDelay = playDelay || 1100;
  this.game = game;
  this.victoryCount = 20;
  this.reset();
}

GameController.prototype = new EventEmitter();

GameController.prototype.getSequenceLength = function getSequenceLength() {
  return this.game.getSequenceLength();
};

GameController.prototype.getStrict = function getStrict() {
  return this.game.getStrict();
};

GameController.prototype.getVictory = function getVictory() {
  return this.game.getSequenceLength() > this.victoryCount;
};

GameController.prototype.getVictoryCount = function getVictory() {
  return this.victoryCount;
};

GameController.prototype.playSequence = function playSequence(sequence) {
  const seq = sequence || this.game.getSequence();

  if (seq.length > 0) {
    let seqIndex = 0;

    this.state = GameState.BUSY;

    const play = () => {
      this.emitEvent('activateGameButton', [seq[seqIndex++], true]);

      if (seq.length > seqIndex) {
        setTimeout(play, this.playDelay);
      } else {
        this.state = GameState.READY;
      }
    };

    play();
  } else {
    throw new Error('Sequence length must be greater than zero.');
  }
};

GameController.prototype.reset = function reset(buttonCount = 8) {
  this.game.reset(buttonCount);
  this.state = GameState.READY;
};

GameController.prototype.submit = function submit(buttonIndex) {
  if (this.getVictory()) {
    return;
  }

  if (this.state === GameState.READY) {
    const lastSequenceLength = this.game.getSequence().length;
    if (this.game.submit(buttonIndex)) {
      this.emitEvent('activateGameButton', [buttonIndex, true]);
      if (lastSequenceLength !== this.game.getSequence().length) {
        //  Don't play sequence after winning
        if (!this.getVictory()) {
          setTimeout(() => this.playSequence(), this.playDelay);
        }
      }
    } else {
      this.emitEvent('activateGameButton', [buttonIndex, false]);
      this.state = GameState.BUSY;
      setTimeout(() => this.playSequence(), this.playDelay);
    }
  }
};

GameController.prototype.toggleStrict = function toggleStrict() {
  this.game.setStrict(!this.game.getStrict());
};

export default GameController;
