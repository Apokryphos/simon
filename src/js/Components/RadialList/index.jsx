import React from 'react';
import PropTypes from 'prop-types';
import GameButton from '../GameButton';
import GameController from '../../GameController';

function RadialList(props) {
  const buttons = [];
  for (let b = 0; b < props.buttonCount; ++b) {
    //  Just use index as key: button list is static and doesn't change
    const key = b;
    buttons.push(<GameButton
      key={key}
      buttonIndex={b}
      gameController={props.gameController}
      onClick={props.onClick}
    />);
  }

  return (
    <div className="radial-list">
      {buttons}
    </div>
  );
}

RadialList.propTypes = {
  buttonCount: PropTypes.number,
  gameController: PropTypes.instanceOf(GameController),
  onClick: PropTypes.func,
};

RadialList.defaultProps = {
  buttonCount: 8,
  gameController: null,
  onClick: null,
};

export default RadialList;
