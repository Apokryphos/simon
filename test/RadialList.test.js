import React from 'react';
import { shallow } from 'enzyme';
import Game from 'Game';
import GameController from 'GameController';
import RadialList from 'Components/RadialList';

describe('RadialList rendering', () => {
  it('Renders OK', () => {
    const wrapper = shallow(<RadialList
      gameController={
          new GameController(new Game(4, () => 0))
        }
      onClick={() => {}}
    />);
    expect(wrapper).toMatchSnapshot();
  });
});
