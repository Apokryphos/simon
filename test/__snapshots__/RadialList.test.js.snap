// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadialList rendering Renders OK 1`] = `
<div
  className="radial-list"
>
  <GameButton
    buttonIndex={0}
    gameController={
      t {
        "game": Game {
          "buttonCount": 8,
          "getRandomButton": [Function],
          "sequence": Array [
            0,
          ],
          "sequenceIndex": 0,
          "strict": false,
        },
        "playDelay": 1100,
        "state": 0,
        "victoryCount": 20,
      }
    }
    key="0"
    onClick={[Function]}
  />
  <GameButton
    buttonIndex={1}
    gameController={
      t {
        "game": Game {
          "buttonCount": 8,
          "getRandomButton": [Function],
          "sequence": Array [
            0,
          ],
          "sequenceIndex": 0,
          "strict": false,
        },
        "playDelay": 1100,
        "state": 0,
        "victoryCount": 20,
      }
    }
    key="1"
    onClick={[Function]}
  />
  <GameButton
    buttonIndex={2}
    gameController={
      t {
        "game": Game {
          "buttonCount": 8,
          "getRandomButton": [Function],
          "sequence": Array [
            0,
          ],
          "sequenceIndex": 0,
          "strict": false,
        },
        "playDelay": 1100,
        "state": 0,
        "victoryCount": 20,
      }
    }
    key="2"
    onClick={[Function]}
  />
  <GameButton
    buttonIndex={3}
    gameController={
      t {
        "game": Game {
          "buttonCount": 8,
          "getRandomButton": [Function],
          "sequence": Array [
            0,
          ],
          "sequenceIndex": 0,
          "strict": false,
        },
        "playDelay": 1100,
        "state": 0,
        "victoryCount": 20,
      }
    }
    key="3"
    onClick={[Function]}
  />
  <GameButton
    buttonIndex={4}
    gameController={
      t {
        "game": Game {
          "buttonCount": 8,
          "getRandomButton": [Function],
          "sequence": Array [
            0,
          ],
          "sequenceIndex": 0,
          "strict": false,
        },
        "playDelay": 1100,
        "state": 0,
        "victoryCount": 20,
      }
    }
    key="4"
    onClick={[Function]}
  />
  <GameButton
    buttonIndex={5}
    gameController={
      t {
        "game": Game {
          "buttonCount": 8,
          "getRandomButton": [Function],
          "sequence": Array [
            0,
          ],
          "sequenceIndex": 0,
          "strict": false,
        },
        "playDelay": 1100,
        "state": 0,
        "victoryCount": 20,
      }
    }
    key="5"
    onClick={[Function]}
  />
  <GameButton
    buttonIndex={6}
    gameController={
      t {
        "game": Game {
          "buttonCount": 8,
          "getRandomButton": [Function],
          "sequence": Array [
            0,
          ],
          "sequenceIndex": 0,
          "strict": false,
        },
        "playDelay": 1100,
        "state": 0,
        "victoryCount": 20,
      }
    }
    key="6"
    onClick={[Function]}
  />
  <GameButton
    buttonIndex={7}
    gameController={
      t {
        "game": Game {
          "buttonCount": 8,
          "getRandomButton": [Function],
          "sequence": Array [
            0,
          ],
          "sequenceIndex": 0,
          "strict": false,
        },
        "playDelay": 1100,
        "state": 0,
        "victoryCount": 20,
      }
    }
    key="7"
    onClick={[Function]}
  />
</div>
`;
